### PR TITLE
Canvas race conditions

### PR DESCRIPTION
**Pull request by dehli**
# Problem 

Currently, there are many images that are required to render the arena. This combines them all into one sprite sheet so that race conditions for image loading is eliminated. 

I also reintroduce image rotation/flipping so we can reuse as many images as possible (reducing spritesheet payload).

Fixes [API #205](https://github.com/willowtreeapps/wombats-api/issues/205)

### CODE REVIEW

File: resources/public/index.html [hunk 1]
```diff
@@ -9,5 +9,6 @@
     <div id="app"></div>
     <script src="js/compiled/app.js"></script>
     <script>wombats_web_client.core.init();</script>
+    <img src="images/spritesheet.png" style="display:none" />
```
Comment (dehli) [inline resources/public/index.html]: I added this image to `index.html` so it loads right when the page loads (that way when the game starts there is minimal wait time).

File: src/cljs/wombats_web_client/events/spritesheet.cljs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+(re-frame/reg-event-db
+ :update-spritesheet
+ (fn [db [_ spritesheet]]
+   (assoc db :spritesheet (:frames spritesheet))))
+
+(re-frame/reg-event-fx
+ :get-spritesheet
+ (fn [_] 
+   (get-spritesheet)
+   nil))
```
Comment (dehli) [inline src/cljs/wombats_web_client/events/spritesheet.cljs]: I was wondering if you knew why the following code doesn't work:

```
(fn [_]
  (get-spriteet))
```

Comment (emilyseibert) [inline src/cljs/wombats_web_client/events/spritesheet.cljs]: https://github.com/Day8/re-frame/wiki/Effectful-Event-Handlers#meet-fx did you happen to read this?

Comment (emilyseibert) [inline src/cljs/wombats_web_client/events/spritesheet.cljs]: try returning a keyword with the function if you want to use `fx`

Comment (emilyseibert) [inline src/cljs/wombats_web_client/events/spritesheet.cljs]: i'm not sure i like get-spritesheet as a separate dispatch though, why wouldn't you just call get-spritesheet?

File: src/cljs/wombats_web_client/components/arena.cljs [hunk 3]
```diff
@@ -2,53 +2,120 @@
...
-                                                  x
-                                                  y
-                                                  width
-                                                  height)))))
-      (set! (.-src img) url))))
+  [canvas-element img-name x y width height]
+  (let [spritesheet (re-frame/subscribe [:spritesheet])
+        sprite-info (get @spritesheet (keyword img-name))
+        frame (:frame sprite-info)]
+    (when-not (nil? frame)
```
Comment (emilyseibert) [inline src/cljs/wombats_web_client/components/arena.cljs]: `(when frame`

File: src/cljs/wombats_web_client/components/arena.cljs [hunk 4]
```diff
@@ -2,53 +2,120 @@
...
+                                                     y
+                                                     width
+                                                     height)))))))
+
+(defn- draw-image-rotated
+  [canvas-element img-name x y width height rotation]
+  (let [spritesheet (re-frame/subscribe [:spritesheet])
+        sprite-info (get @spritesheet (keyword img-name))
+        frame (:frame sprite-info)]
+    (when-not (nil? frame)
```
Comment (emilyseibert) [inline src/cljs/wombats_web_client/components/arena.cljs]: same here

File: src/cljs/wombats_web_client/components/arena.cljs [hunk 5]
```diff
@@ -2,53 +2,120 @@
...
+                                                             width
+                                                             height
+                                                             rotation)))))))
+
+(defn- draw-image-flipped-horizontally
+  [canvas-element img-name x y width height]
+  (let [spritesheet (re-frame/subscribe [:spritesheet])
+        sprite-info (get @spritesheet (keyword img-name))
+        frame (:frame sprite-info)]
+    (when-not (nil? frame)
```
Comment (emilyseibert) [inline src/cljs/wombats_web_client/components/arena.cljs]: hi

File: src/cljs/wombats_web_client/components/arena.cljs [hunk 6]
```diff
@@ -2,53 +2,120 @@
...
+                                                             (:w frame)
+                                                             (:h frame)
+                                                             x y
+                                                             width
+                                                             height
+                                                             rotation)))))))
+
+(defn- draw-image-flipped-horizontally
+  [canvas-element img-name x y width height]
+  (let [spritesheet (re-frame/subscribe [:spritesheet])
```
Comment (emilyseibert) [inline src/cljs/wombats_web_client/components/arena.cljs]: this logic is repeated for draw -images..can we combine somehow?

File: src/cljs/wombats_web_client/events/spritesheet.cljs (hunk 2 above)
Comment (dehli) [inline src/cljs/wombats_web_client/events/spritesheet.cljs]: Nope! I'll read through that. Thanks! And that sounds like a better way to do it.

File: src/cljs/wombats_web_client/components/arena.cljs (hunk 6 above)
Comment (dehli) [inline src/cljs/wombats_web_client/components/arena.cljs]: Yaaa, meant to come back and clean that up but forgot. Thanks!

File: src/cljs/wombats_web_client/components/arena.cljs (hunk 5 above)
Comment (dehli) [inline src/cljs/wombats_web_client/components/arena.cljs]: 👋

File: src/cljs/wombats_web_client/events/spritesheet.cljs (hunk 2 above)
Comment (emilyseibert) [inline src/cljs/wombats_web_client/events/spritesheet.cljs]: i use `event-fx` to chain dispatches and db updates when i bootstrap the app post login if you want to see how it's used in the app currently.

**Comment by dehli** [comment]
Updated based on your great 🍔 🔙  @emilyseibert. Thanks!

**Comment by emilyseibert** [comment]
Nicely done!

**Pull request merged**